app/uploaders/voice_uploader.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+class VoiceUploader < CarrierWave::Uploader::Base
```

Comment:
I think defining a new ```VoiceUploader``` is necessary, since we only accept voice file, i.e. ```wav``` and ```mp3``` files. Moreover, this voice files may need to be handled differently in the future. E.g. Maybe in the future we want to compress the voice, or even add voice changer function to normal voice files.

Reply:
You can still do the processing/white list extension in the answers model without the uploader. 

The uploader way will work too, but it's hard for us to track files on S3 and it's inconsistent with the implementations of the rest Coursemology code. I think this is one tradeoff that we have to make.

What do you think?

Reply:
I would prefer to keep this Uploader. The functionalities of the Uploader is to handle and process the files. Many APIs, such as whitelist, file version, etc, have already been built in this library. Since ```Carrierwave``` has already provided us these APIs, by convention we should use them directly. It is not good to reinvent the wheel, and handle these again in the model. And I don't think these functionalities are easy to implement again in the model.

In my understanding, the ```Attachment``` is a generic file uploading method. If we do not need to do any special process to the file, we may use it. It is just for storing and downloading. The system do not even need to care about the file type in this case.

But if file need to upload is not just for storing and downloading, and the system need to process it in some special way, we should use a special ```Uploader``` to handle this type of file. This is why we added other uploaders, such as ```ImageUploader``` and ```VoiceUploader```.

If you really think it is hard to track the file in S3, do you think we need to change the ```store_dir``` in the ```VoiceUploader``` instead?

Reply:
> And I don't think these functionalities are easy to implement again in the model.

It won't be hard as the extension whitelist can be done using model validations. And the complexity of the processing will go the the actual processing part. 

> In my understanding, the Attachment is a generic file uploading method. If we do not need to do any special process to the file, we may use it. It is just for storing and downloading. The system do not even need to care about the file type in this case.

Attachment framework is created to prevent duplicate files, and tracking files in S3. the `store_dir` won't help because the deleted files are not deleted in S3 and thus become untracked. It's an abstraction layer between Carrierwave and Coursemology, and we are lean to build most of the things on top of it, we could likely move image uploader to it in future.

But I also heard you, this voice question probably won't have duplications. I think if you insists and want to go with this approach, just make sure that the files are deleted on S3 if the model is deleted.

Reply:
Ok, thanks. I will keep the ```Uploader``` instead.

Reply:
@huangjosh maybe you can try this, does it work? https://stackoverflow.com/questions/36876341/use-different-carrierwave-uploaders-for-one-model-based-on-attributes-of-another

I will read through later and give my views if necessary!